lib/instrumentation/trace-context.js

```diff
@@ -0,0 +1,162 @@
+      return value
+    }
+  })
+}
+
+function hexSliceFn (buffer, offset, length) {
+  return () => buffer.slice(offset, length).toString('hex')
+}
+
+function maybeHexSliceFn (buffer, offset, length) {
+  const fn = hexSliceFn(buffer, offset, length)
+  return () => {
+    const value = fn()
+    // Check for any non-zero characters to identify a valid ID
+    if (/[1-9a-f]/.test(value)) {
+      return value
+    }
+  }
+}
+
+function makeChild (buffer) {
+  // Move current id into parentId region
+  buffer.copy(buffer, OFFSETS.parentId, OFFSETS.id, OFFSETS.flags)
+
+  // If the parent context has been requested,
+  // mark the child context as reported.
+  if (buffer[OFFSETS.flags] & FLAGS.requested) {
+    buffer[OFFSETS.flags] |= FLAGS.reported
+  } else {
+    buffer[OFFSETS.flags] &= ~FLAGS.reported
+  }
+
+  // Generate new id
+  crypto.randomFillSync(buffer, OFFSETS.id, SIZES.id)
+
+  return new TraceContext(buffer)
+}
+
+function isValidHeader (header) {
+  return /^[\da-f]{2}-[\da-f]{32}-[\da-f]{16}-[\da-f]{2}$/.test(header)
+}
+
+// NOTE: The version byte is not fully supported yet, but is not important until
+// we use the official header name rather than Elastic-APM-traceparent.
+// https://w3c.github.io/distributed-tracing/report-trace-context.html#versioning-of-traceparent
+function headerToBuffer (header) {
+  const buffer = Buffer.alloc(SIZES.all)
+  buffer.write(header.replace(/-/g, ''), 'hex')
+  return buffer
+}
+
+function resume (header) {
+  return makeChild(headerToBuffer(header))
+}
+
+function start (sampled = false) {
+  const buffer = Buffer.alloc(SIZES.all)
+
+  // Generate new ids
+  crypto.randomFillSync(buffer, OFFSETS.traceId, SIZES.ids)
+
+  // Apply reported and requested flags
+  if (sampled) {
+    buffer[OFFSETS.flags] |= FLAGS.reported
+    buffer[OFFSETS.flags] |= FLAGS.requested
+  }
+
+  return new TraceContext(buffer)
+}
+
+const bufferSymbol = Symbol('trace-context-buffer')
+
+class TraceContext {
+  constructor (buffer) {
+    this[bufferSymbol] = buffer
+    defineLazyProp(this, 'version', hexSliceFn(buffer, OFFSETS.version, OFFSETS.traceId))
+    defineLazyProp(this, 'traceId', hexSliceFn(buffer, OFFSETS.traceId, OFFSETS.id))
+    defineLazyProp(this, 'id', hexSliceFn(buffer, OFFSETS.id, OFFSETS.flags))
+    defineLazyProp(this, 'flags', hexSliceFn(buffer, OFFSETS.flags, OFFSETS.parentId))
+    defineLazyProp(this, 'parentId', maybeHexSliceFn(buffer, OFFSETS.parentId))
+
+    const requested = !!(this[bufferSymbol][OFFSETS.flags] & FLAGS.requested)
+    const reported = !!(this[bufferSymbol][OFFSETS.flags] & FLAGS.reported)
+    Object.defineProperties(this, {
+      requested: {
+        enumerable: true,
+        value: requested
+      },
+      reported: {
+        enumerable: true,
+        value: reported
+      },
+      sampled: {
+        enumerable: true,
+        value: reported
+      }
+    })
+  }
+
+  static startOrResume (traceparent, conf) {
+    return isValidHeader(traceparent)
```

Comment:
A potential performance improvement would be to return a boolean from resume which indicates whether the header could be parsed without errors. If it's false, start a new trace. That way, you can get rid of the regex check.

Just an idea, feel free to ignore :)

Reply:
The parsing can't fail. It just copies whatever is in the header into the buffer memory, regardless of what is there. The regex check needs to be made _somewhere_ to validate the contents of the data to be copied into the buffer, might as well double as a switch between resuming from a header or starting a new context.

Reply:
See also https://github.com/elastic/apm-agent-java/blob/d8b583f30c6b406ba8f81e63f0ce71af22dc2469/apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/Transaction.java#L85-L87

Reply:
Sorry, didn't see your previous comment. I implemented the hex parsing in a way that it throws an exception if there are invalid chars: https://github.com/felixbarny/apm-agent-java/blob/e78d187659ea77899c6f928de93f52ec0028ced6/apm-agent-core/src/main/java/co/elastic/apm/util/HexUtils.java#L67.
That's why with this approach, there is no need to validate upfront. In case I did not overlook something, that is.